File: samples/typescript/src/orderRepo.ts
// Persistence for orders. In-memory here.

import { Money } from "./money";
import { Order } from "./models";

export class OrderRepository {
  private orders = new Map<string, Order>();
  private sequence = 0;

  nextId(): string {
    this.sequence += 1;
    return `ORD-${String(this.sequence).padStart(6, "0")}`;
  }

  save(customerId: string, total: Money): Order {
    const order: Order = {
      orderId: this.nextId(), customerId, total, status: "pending",
    };
    this.orders.set(order.orderId, order);
    return order;
  }

  get(orderId: string): Order | undefined {
    return this.orders.get(orderId);
  }

  markPaid(orderId: string): Order {
    const order = this.orders.get(orderId);
    if (!order) throw new Error(`unknown order ${orderId}`);
    order.status = "paid";
    return order;
  }
}

File: samples/rust/src/order_repo.rs
//! Persistence for orders. In-memory here.

use std::collections::HashMap;

use crate::models::Order;
use crate::money::Money;

/// Stores and retrieves orders by id.
pub struct OrderRepository {
    orders: HashMap<String, Order>,
    sequence: u32,
}

impl OrderRepository {
    /// Builds an empty repository.
    pub fn new() -> OrderRepository {
        OrderRepository { orders: HashMap::new(), sequence: 0 }
    }

    /// Allocates a stable, sortable order id.
    pub fn next_id(&mut self) -> String {
        self.sequence += 1;
        format!("ORD-{:06}", self.sequence)
    }

    /// Persists a new order and returns a copy.
    pub fn save(&mut self, customer_id: &str, total: Money) -> Order {
        let order = Order {
            order_id: self.next_id(),
            customer_id: customer_id.to_string(),
            total,
            status: "pending".to_string(),
        };
        self.orders.insert(order.order_id.clone(), order.clone());
        order
    }

    /// Returns an order if it is known.
    pub fn get(&self, order_id: &str) -> Option<&Order> {
        self.orders.get(order_id)
    }

    /// Moves an order to the paid state.
    pub fn mark_paid(&mut self, order_id: &str) -> Option<Order> {
        let order = self.orders.get_mut(order_id)?;
        order.status = "paid".to_string();
        Some(order.clone())
    }
}

impl Default for OrderRepository {
    fn default() -> Self {
        Self::new()
    }
}
